# Imports

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# filepath for customers.csv

customers_filepath = '/Users/daphne/Downloads/Instacart Basket Analysis/02 Data/Original Data/customers.csv'


In [3]:
# Open and read file

df_cust = pd.read_csv(customers_filepath)

In [29]:
import pickle

In [30]:
# filepath for ords_prods_merge_V2.pkl

ords_prods_filepath = '/Users/daphne/Downloads/Instacart Basket Analysis/02 Data/Prepared Data/ords_prods_merge_V2.pkl'


In [31]:
# Open and read the Pickle file


with open(ords_prods_filepath, 'rb') as file:
    ords_prods_merge = pickle.load(file)

# Task

## Step 4

In [4]:
# observing dataframe

df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
# observing dataframe

df_cust.shape

(206209, 10)

Changing column names because it is confusing.

In [8]:
# renaming "First Name" column to first_name

df_cust.rename(columns={'First Name': 'first_name'}, inplace=True)

In [9]:
# renaming "Surnam" column to first_name

df_cust.rename(columns={'Surnam': 'surname'}, inplace=True)

In [10]:
# renaming "Gender" column to first_name

df_cust.rename(columns={'Gender': 'gender'}, inplace=True)

In [11]:
# renaming "STATE" column to first_name

df_cust.rename(columns={'STATE': 'state'}, inplace=True)

In [12]:
# renaming "Age" column to first_name

df_cust.rename(columns={'Age': 'age'}, inplace=True)

In [13]:
df_cust.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [14]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    194950 non-null  object
 2   surname       206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## Step 5

In [15]:
# Checking for missing values

df_cust.isnull().sum()

user_id             0
first_name      11259
surname             0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [16]:
# Looking at missing values

df_nan = df_cust[df_cust['first_name'].isnull() == True]

In [ ]:
df_nan.head(50)

In [ ]:
df_nan.tail(50)

There is no obvious pattern with the 11,259 missing values. This will be left as is.

In [25]:
# Looking for duplicates

df_dups = df_cust[df_cust.duplicated()]

In [26]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

No Duplicates

In [27]:
# Looking for mixed type data

for col in df_cust.columns.tolist():
  weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cust[weird]) > 0:
    print (col)

first_name


In [28]:
# fixing mixed type data in first_name column

df_cust['first_name'] = df_cust['first_name'].astype('str')

## Step 6

In [32]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 26 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   order_dow               int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  _merge                  category
 11  product_name            object  
 12  aisle_id                int64   
 13  department_id           int64   
 14  prices                  float64 
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  mean_p

In [33]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   first_name    206209 non-null  object
 2   surname       206209 non-null  object
 3   gender        206209 non-null  object
 4   state         206209 non-null  object
 5   age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [34]:
# Drop _merge column from ords_prods_merge

ords_prods_merge = ords_prods_merge.drop('_merge', axis=1)

In [35]:
# Merge dataframes on the 'user_id' column

merged_df = pd.merge(ords_prods_merge, df_cust, on='user_id', how='inner')

In [ ]:
merged_df.head(20)

## Step 8

In [37]:
# Defining filepath

file_path = "/Users/daphne/Downloads/Instacart Basket Analysis/02 Data/instacart_basket_df.pkl"

In [38]:
# Export dataframe

merged_df.to_pickle(file_path)

In [39]:
# Comfirm export

print(f"DataFrame exported to '{file_path}' successfully.")

DataFrame exported to '/Users/daphne/Downloads/Instacart Basket Analysis/02 Data/instacart_basket_df.pkl' successfully.
